### PR TITLE
Shard MHA. 

### PR DESCRIPTION
**Pull request by wujingyue**
As a follow-up to #3045. 

For #2199.

### CODE REVIEW

**Comment by wujingyue** [comment]
!build

**Comment by wujingyue** [comment]
!build

File: tests/python/test_multidevice.py [hunk 1]
```diff
@@ -332,35 +332,49 @@ def definition(self) -> None:
...
+        #                   [d,b,s,e/d]        [d,e,e/d]                 [e]
+        T152_local_matmul = self.ops.matmul(
+            T151,
+            self.ops.broadcast_in_dim(
+                self.ops.permute(self.mha_linear1_weight, [0, 2, 1]),
+                [d, 1, e // d, e],
+                [0, 2, 3],
+            ),
+        )
+        T152_matmul = self.ops.sum(T152_local_matmul, [0])  # allreduce
```
Comment (samnordmann) [inline tests/python/test_multidevice.py:372]: I wonder what would happen currently if we do not decompose the matmul and the allreduce...

File: tests/python/test_multidevice.py [hunk 2]
```diff
@@ -332,35 +332,49 @@ def definition(self) -> None:
...
+            start_indices=[0, 0, 0, e * 2 // d],
+            end_indices=[d, b, s, e * 3 // d],
         )
-        T123 = self.ops.reshape(T104, new_shape=[b, s, h, e // h])
-        T124 = self.ops.permute(T123, dims=[0, 2, 1, 3])
-        T130 = self.ops.reshape(T91, new_shape=[b, s, h, e // h])
-        T131 = self.ops.permute(T130, dims=[0, 2, 1, 3])
-        T137 = self.ops.reshape(T117, new_shape=[b, s, h, e // h])
-        T138 = self.ops.permute(T137, dims=[0, 2, 1, 3])
+        T123 = self.ops.reshape(T104, new_shape=[d, b, s, h // d, e // h])
```
Comment (samnordmann) [inline tests/python/test_multidevice.py:348]: IIUC, we pass from shape `[d, b, s, e//d]` to `[d, b, s, h//d, e//h]`. Nothing illegal about it but it looks surprising to me so I just want to make sure

**Review by samnordmann — APPROVED** [review]
LGTM

File: tests/python/test_multidevice.py (hunk 2 above)
Comment (wujingyue) [reply tests/python/test_multidevice.py:348]: I double checked -- it looks right. MHA is head parallel according to Figure 3b in https://arxiv.org/pdf/1909.08053.

File: tests/python/test_multidevice.py (hunk 1 above)
Comment (wujingyue) [reply tests/python/test_multidevice.py:372]: The first thing that'll break is that `linear` will produce a wrong shape. linear, as is implemented today, will output a tensor of rank `input_rank + weight_rank - 2 = 5`. However, we want the shape to be `[d,b,s,e]` and thus 4D.

**Comment by wujingyue** [comment]
!build

**Pull request merged**